### PR TITLE
TRD trapsimulator correctly produce raw data.

### PR DESCRIPTION
**Pull request by bazinski**
- Change to 64 tracklet output of trapsimulator.
- Correctly implement raw data out of trapsimulator.
- halfchamber header in a separate link record akin to the way the hardware actually works. 
- Add additional output of TrapSimulator to TRDTrackRawWriterSpec.
- Some other sundry fixups to sort all the above out.
- Header data structures for raw data names changed

### CODE REVIEW

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h [hunk 1]
```diff
@@ -207,14 +209,15 @@ struct TrapRawTracklet { // This is a bad name as part of the tracklet data is i
   union {
     uint32_t word;
     struct {
-      uint8_t checkbit : 1; // The size of the data for this link
+      uint8_t checkbit : 1; // 
       uint16_t pid : 15;    // Particle Identity
       uint16_t slope : 6;   // Deflection angle of tracklet
       uint16_t pos : 10;    // Position of tracklet, signed 10 bits, granularity 0.02 pad widths, -10.22 to +10.22, relative to centre of pad 10
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h:215]: I think this comment is not up to date.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h [hunk 2]
```diff
@@ -276,160 +279,25 @@ uint32_t getSlope(const *rawdata, int tracklet)
...
-  // make a pretty output of the tracklet.
-  stream << "TrapRawTracklet: pos=" << tracklet.pos << "::slope=" << tracklet.slope << "::pid=" << tracklet.pid << "::checkbit=" << tracklet.checkbit << std::endl;
-  return stream;
-}
-void printTrapRawTracklet(o2::trd::TrapRawTracklet& tracklet)
-{
-  LOGF(INFO, "TrapRawTracklet: ");
-}
 
-void printMCMHeader(o2::trd::MCMRawDataHeader& mcmhead)
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h:369]: I just spotted an issue here: you are actually printing a tracklet header. The name MCMRawDataHeader would more intuitively be reserved for the digit/ADC data headers.
Should we change this now, or do a review of the raw data struct in the next week or two?

File: DataFormats/Detectors/TRD/src/RawData.cxx [hunk 3]
```diff
@@ -9,13 +9,170 @@
...
 namespace o2
 {
 
 namespace trd
 {
 
+uint32_t unpacklinkinfo(const HalfCRUHeader& cruhead, const uint32_t link, const bool data = true)
+{
+    //TODO this needs some more ellaborate testing. I got the packing to work so for now,
+    //     its a question of unpacking info in line 88 and 89.
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/src/RawData.cxx:23]: This should also be documented, i.e. what goes into the uint32_t and at what bit positions. I

**Review by tdietel — DISMISSED** [review]
Looks ok for now, but we should review the naming of the raw data structs, and check for redundant print functions / ostream operators.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h (hunk 2 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h:369]: Now would be the best time to change it yes. We can discuss in the current meeting.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h (hunk 1 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h:215]: mmmm yes. pid and slope must swap around

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h [hunk 4]
```diff
@@ -276,160 +279,25 @@ uint32_t getSlope(const *rawdata, int tracklet)
...
-  LOGF(INFO, "MCMRawHeader: 1:%d padrow: 0x%02x col: 0x%01x pid2 0x%02x pid1: 0x%02x pid0: 0x%02x 1:%d", mcmhead.onea, mcmhead.padrow, mcmhead.col, mcmhead.pid2, mcmhead.pid1, mcmhead.pid0, mcmhead.oneb);
-}
 
-std::ostream& operator<<(std::ostream& stream, const MCMRawDataHeader& mcmhead)
-{
-  // make a pretty output of the mcm header.
-  // stream << "MCMRawHeader:" << mcmhead.checkbits << "::" << (mcmhead.pid&0xfff000000) << ":"<<  (mcmhead.pid&0xfff000) << ":"<< (mcmhead.pid&0xfff) << std::endl;
-  stream << "MCMRawHeader:" << mcmhead.onea << "::" << mcmhead.pid2 << ":" << mcmhead.pid1 << ":" << mcmhead.pid0 << "::" << mcmhead.oneb << std::endl;
-  return stream;
-}
```
[the remark this answers is not in the document]
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h:380]: they are the same just different ways of calling the same thing.

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h (hunk 1 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/RawData.h:214]: fixed in 68107d8d41c

File: DataFormats/Detectors/TRD/src/RawData.cxx (hunk 3 above)
Comment (bazinski) [reply DataFormats/Detectors/TRD/src/RawData.cxx]: fixed in 7dac991d840

**Comment by bazinski** [comment]
Sorry about the number of files, they are all side effects from making the trapsimulator output what actually comes out of the chips and arrives at the cru.

**Comment by bazinski** [comment]
@tdietel could approve this so I can get it merged? I have 2 more PRs that are based off this.

**Comment by tdietel** [comment]
Hi Sean.

I think some of the class/struct names might still be ambiguous when we consider the digits data that still has to come in. 

HC headers will be there in the new format for tracklets, and in the old format for the digits. Similar for MCMDataHeader. I would suggest to rename them:

- HalfChamberHeader -> TrackletHCheader
- MCMDataHeader -> TrackletMCMheader
- MCMDataHeader -> TrackletData or TrackletPayload or TrackletMCMdata

I don't mind the exact names, but I think it should be clear that these are related to the tracklet data and not to the digits. 

Cheers,
 - Tom.

**Comment by bazinski** [comment]
Hi @tdietel 
Names changed, the ending of header/data is capitalised.

-  TrackletMCMHeader
- TrackletMCMHeader
- TrackletMCMData

**Review by tdietel — DISMISSED** [review]
All comments have been addressed. I approve the PR.

**Comment by tdietel** [comment]
+1

**Review by tdietel — DISMISSED** [review]

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h [hunk 5]
```diff
@@ -30,8 +30,7 @@
 #include <memory>   // for std::unique_ptr
 #include "Rtypes.h" // for ClassDef
 #include "fairlogger/Logger.h"
-#include "TRDBase/FeeParam.h"
-
+#include <iostream>
```
Comment (sawenzel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h:33]: If possible avoid inclusion of iostream in fundamental headers. Use <iosfwd>

Comment (bazinski) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h]: Removed it and numerous headers from RawData.h. Thanks for the catch.

**Pull request merged**